### PR TITLE
Balder/add attribute operation

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe  or @toregge PR 
@geirst FYI
This is a standalone part of #6609 extended with tests

### CODE REVIEW

**Review by havardpe — COMMENTED** [review]
I think this code should be co-located with the attribute code (low-level type dependencies), not with matching.

File: searchcore/src/tests/proton/matching/attribute_operation_test.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+TEST("test all float operations") {
+    auto attr = createAttribute(BasicType::DOUBLE, "af");
+    const std::vector<std::pair<const char *, double>> expectedOperation = {
+            {"++", 8}, {"--", 6}, {"+=7.3", 14.3}, {"-=0.9", 6.1}, {"*=3.1", 21.7}, {"/=2", 3.5}, {"%=3", 7}
+    };
+    for (auto operation : expectedOperation) {
+        TEST_DO(verify<double>(operation.first, *attr, 7, operation.second));
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/tests/proton/matching/attribute_operation_test.cpp]: Done

File: searchcore/src/vespa/searchcore/proton/matching/attribute_operation.cpp [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+                          [&op](uint32_t docId) { op(docId); });
+        }
+    }
+private:
+    typename OP::F::V _operand;
+    std::vector<uint32_t> _docIds;
+};
+
+struct Operation {
+    enum Type { INC, DEC, ADD, SUB, MUL, DIV, MOD, SET, BAD };
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/matching/attribute_operation.cpp]: Done

File: searchcore/src/vespa/searchcore/proton/matching/attribute_operation.cpp [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+            size_t             numHits   = _result->getArrayUsed();
+            std::for_each(hits, hits+numHits,  [&op](search::RankedHit hit) { op(hit.getDocId()); });
+            if (_result->getBitOverflow()) {
+                _result->getBitOverflow()->foreach_truebit([&op](uint32_t docId) { op(docId); });
+            }
+        }
+    }
+private:
+    typename OP::F::V _operand;
+    std::unique_ptr<search::ResultSet> _result;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/matching/attribute_operation.cpp]: Fixed, by using a pair of a bitvector and a vector.

**Comment by baldersheim** [comment]
@havardpe I think this one can go in now.

File: searchlib/src/vespa/searchlib/common/resultset.cpp [hunk 4]
```diff
@@ -69,54 +58,53 @@ ResultSet::mergeWithBitOverflow(HitRank default_value)
...
     setBitOverflow(nullptr);
 }
 
+void
+ResultSet::sort(FastS_IResultSorter & sorter, unsigned int ntop) {
+    sorter.sortResults(&_rankedHitsArray[0], _rankedHitsArray.size(), ntop);
+}
+
+std::pair<std::unique_ptr<BitVector>, vespalib::Array<RankedHit>>
+ResultSet::move(ResultSet && rhs) {
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/common/resultset.cpp]: I would prefer something else than 'move' here, maybe 'steal_internals' or 'detach' or something...

**Comment by havardpe** [comment]
The code itself looks good
I am more concerned about the implications of enabling this (security, caching, mutability)

**Comment by bratseth** [comment]
We also need work on the container level to make this a usable feature for end users. That will come later.

As extra safety I think the backend should also refuse to mutate attributes that have a document field as source.

**Comment by baldersheim** [comment]
I thought we might want to explicit note the attribute as 'mutable'.

**Comment by bratseth** [comment]
Good idea, but it shouldn't be possible for attributes of document fields.
If we do it explicit it could make sense also for fields outside the document which is sourced from the document.

**Comment by baldersheim** [comment]
Agreed.

**Comment by baldersheim** [comment]
@geirst FYI

**Pull request merged**